File: Week 1/credit.c
#include <cs50.h>
#include <stdio.h>

string verifyCreditCard(long n);
int getDigits(long n);
bool luhnVerification(long creditCard);
int main(void)
{
    long creditCard;

    creditCard = get_long("Credit Card: ");
    printf("%s\n", verifyCreditCard(creditCard));
}

string verifyCreditCard(long n)
{
    int digits = getDigits(n);

    if (digits != 13 && digits != 15 && digits != 16)
    {
        return "INVALID";
    }

    if (!luhnVerification(n))
    {
        return "INVALID";
    }

    if (digits == 13)
    {
        if (n / 1000000000000 == 4)
        {
            return "VISA";
        }
    }

    if (digits == 15)
    {
        if (n / 10000000000000 == 37 || n / 10000000000000 == 34)
        {
            return "AMEX";
        }
    }

    if (digits == 16)
    {
        printf("entrou 16");
        if (n / 1000000000000000 == 4)
        {
            return "VISA";
        }

        if (n / 100000000000000 == 51 || n / 100000000000000 == 52 || n / 100000000000000 == 53 ||
            n / 100000000000000 == 54 || n / 100000000000000 == 55)
        {
            return "MASTERCARD";
        }
    }

    return "INVALID";
}

int getDigits(long n)
{
    int digits = 0;
    if (n == 0)
    {
        return 1;
    }

    while (n > 0)
    {
        n = n / 10;
        digits++;
    }
    return digits;
}

bool luhnVerification(long creditCard)
{
    int doubled = 0;
    int others = 0;
    long remainingDigits = creditCard;
    int digits = getDigits(creditCard);

    bool shouldDouble = false;

    for (int i = 0; i < digits; i++)
    {
        int lastDigit = remainingDigits % 10;
        remainingDigits = remainingDigits / 10;
        if (shouldDouble)
        {
            if (lastDigit * 2 >= 10)
            {
                doubled += (lastDigit * 2) % 10;
                doubled += (lastDigit * 2) / 10;
            }
            else
            {
                doubled += (lastDigit * 2);
            }
        }
        else
        {
            others += lastDigit;
        }
        shouldDouble = !shouldDouble;
    }
    printf("%d\n", doubled);
    printf("%d\n", others);
    int sum = doubled + others;

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

File: Week 6/sentimental-credit/credit.py
from cs50 import get_int
import math


def main():
    creditCard = str(get_int("Credit card: "))
    print(verifyCreditCard(creditCard))


def verifyCreditCard(creditCard):
    digits = getDigits(creditCard)
    hasWrongSize = digits != 13 and digits != 15 and digits != 16

    if hasWrongSize or not luhnVerification(creditCard):
        return "INVALID"

    if digits == 13 and creditCard[0] == '4':
        return "VISA"

    if digits == 15 and creditCard[0] == '3' and (creditCard[1] == '7' or creditCard[1] == '4'):
        return "AMEX"

    if digits == 16:
        if creditCard[0] == '4':
            return "VISA"

        if creditCard[0] == '5' and (int(creditCard[1]) in range(1, 6)):
            return "MASTERCARD"

    return "INVALID"


def getDigits(creditCard):
    return len(creditCard)


def luhnVerification(creditCard):
    doubled = 0
    others = 0
    remainingDigits = creditCard

    shouldDouble = False
    for digit in creditCard:
        lastDigit = int(remainingDigits[-1])
        remainingDigits = remainingDigits[:-1]
        if shouldDouble:
            if lastDigit * 2 >= 10:
                doubled += (lastDigit * 2) % 10
                doubled += math.floor((lastDigit * 2) / 10)
            else:
                doubled += lastDigit * 2
        else:
            others += lastDigit
        shouldDouble = not shouldDouble
    sum = doubled + others
    return True if sum % 10 == 0 else False


main()
